dp/drivers/i40e.c

```diff
@@ -128,10 +158,186 @@ static int reta_update(struct ix_rte_eth_dev *dev, struct rte_eth_rss_reta *reta
+			log_err("i40e: TCP RX checksum error, dropping pkt\n");
+			valid_checksum = false;
+		}
+
+		/* translate descriptor info into mbuf parameters */
+		b = rxqe->mbuf;
+		b->len = ((qword1 & I40E_RXD_QW1_LENGTH_PBUF_MASK) >>
+				I40E_RXD_QW1_LENGTH_PBUF_SHIFT);
+
+		if (qword1 & (1 << I40E_RX_DESC_STATUS_FLM_SHIFT)) {
+			b->fg_id = MBUF_INVALID_FG_ID;
+		} else {
+			local_fg_id = (le32_to_cpu(rxd.wb.qword0.hi_dword.rss) &
+					   (ETH_RSS_RETA_NUM_ENTRIES - 1));
+			b->fg_id = rx->dev->data->rx_fgs[local_fg_id].fg_id;
+		}
+		b->timestamp = timestamp;
+
+		new_b = mbuf_alloc_local();
+		if (unlikely(!new_b)) {
+			log_err("i40e: unable to allocate RX mbuf\n");
+			goto out;
+		}
+
+		maddr = mbuf_get_data_machaddr(new_b);
+		rxqe->mbuf = new_b;
+		rxdp->read.hdr_addr = rte_cpu_to_le_64(maddr);
+		rxdp->read.pkt_addr = rte_cpu_to_le_64(maddr);
+
+		if (unlikely(!valid_checksum || eth_recv(rx, b))) {
+			log_info("i40e: dropping packet\n");
+			mbuf_free(b);
+		}
+
+		rxq->head++;
+		nb_descs++;
+	}
+
+out:
+
+	/*
+	 * We threshold updates to the RX tail register because when it
+	 * is updated too frequently (e.g. when written to on multiple
+	 * cores even through separate queues) PCI performance
+	 * bottlnecks have been observed.
+	 */
+	if ((uint16_t)(rxq->len - (rxq->tail + 1 - rxq->head)) >=
+		I40E_RDT_THRESH) {
+		rxq->tail = rxq->head + rxq->len - 1;
+
+		/* inform HW that more descriptors have become available */
+		I40E_PCI_REG_WRITE(rxq->rdt_reg_addr,
+					(rxq->tail & (rxq->len - 1)));
+	}
+
+	return nb_descs;
+}
+
+/**
+ * rx_queue_setup - prepares an RX queue
+ * @dev: the ethernet device
+ * @queue_idx: the queue number
+ * @numa_node: the desired NUMA affinity, or -1 for no preference
+ * @nb_desc: the number of descriptors to create for the ring
+ *
+ * Returns 0 if successful, otherwise failure.
+ */
 static int rx_queue_setup(struct ix_rte_eth_dev *dev, int queue_idx, int numa_node, uint16_t nb_desc)
 {
-	panic("rx_queue_setup not implemented\n");
+	void *page;
+	machaddr_t page_phys;
+	int ret;
+	struct rx_queue *rxq;
+
+	/*
+	 * The number of receive descriptors must not exceed hardware
+	 * maximum and must be a multiple of I40E_RING_BASE_ALIGN.
+	 */
+	BUILD_ASSERT(align_up(sizeof(struct rx_queue), I40E_RING_BASE_ALIGN) +
+			 align_up(sizeof(union i40e_rx_desc) * I40E_MAX_RING_DESC, I40E_RING_BASE_ALIGN) +
+			 sizeof(struct rx_entry) * I40E_MAX_RING_DESC < PGSIZE_2MB);
+
+	/*
+	 * Additionally, for purely software performance optimization reasons,
+	 * we require the number of descriptors to be a power of 2.
+	 */
+	if (nb_desc & (nb_desc - 1))
+		return -EINVAL;
+
+	/* NOTE: This is a hack, but it's the only way to support late/lazy
+	 * queue setup in DPDK; a feature that IX depends on. */
+	rte_eth_devices[dev->port].data->nb_rx_queues = queue_idx + 1;
+
+	ret = rte_eth_rx_queue_setup(dev->port, queue_idx, nb_desc, numa_node, &rx_conf, dpdk_pool);
```

Comment:
Using global `rx_conf` here. Need to check it works as I accidentally was using a zero-filled default conf (and it worked surprisingly!). We can also use `NULL` and have dpdk use a default one.